src/server.c

```diff
@@ -695,30 +695,30 @@ int allPersistenceDisabled(void) {
 
 /* Add a sample to the operations per second array of samples. This takes total operation count and
  * current time, then calculates and records the operation rate between to samples. */
-void trackInstantaneousRateMetric(int metric, long long current_reading, long long current_time_us) {
+void trackInstantaneousRateMetric(int metric, long long current_reading, monotime current_time) {
     if (server.inst_metric[metric].last_sample_time > 0) {
-        long long time_duration = current_time_us - server.inst_metric[metric].last_sample_time;
+        monotime time_duration = current_time - server.inst_metric[metric].last_sample_time;
         long long ops = current_reading - server.inst_metric[metric].last_sample_count;
         long long ops_per_sec = time_duration > 0 ? (ops * 1000000 / time_duration) : 0;
         server.inst_metric[metric].samples[server.inst_metric[metric].idx] = ops_per_sec;
         server.inst_metric[metric].idx++;
         server.inst_metric[metric].idx %= STATS_METRIC_SAMPLES;
     }
-    server.inst_metric[metric].last_sample_time = current_time_us;
+    server.inst_metric[metric].last_sample_time = current_time;
     server.inst_metric[metric].last_sample_count = current_reading;
 }
 
 /* Add a sample to time per operation array of samples. This takes total operation count and total time 
    consumption, and record the average time consumption of one operation between two samples. */
-void trackInstantaneousAvgMetric(int metric, long long current_sum, long long current_cnt) {
-    long long time = current_sum - server.inst_metric[metric].last_sample_time;
-    long long cnt = current_cnt - server.inst_metric[metric].last_sample_count;
+void trackInstantaneousAvgMetric(int metric, long long current_reading, monotime current_time_sum) {
+    long long time = current_time_sum - server.inst_metric[metric].last_sample_time;
```

Comment:
this isn't really a monotime (timestamp), it's a duration.
we can maybe add a union, but maybe that's excessive.
at least the monotime tells us what are the units.
maybe some comment can improve things a bit.
e.g. `current_time_sum is a the total duration of that metric in microseconds`

Reply:
IMHO, the rate and avg are both average, the only difference is the dividend and divisor, the arguments `current_reading` and `current_time` here are unclear, I prefer rename them and reuse the original `trackInstantaneousMetric` like:

```diff
--- a/src/server.h
+++ b/src/server.h
@@ -1695,8 +1695,8 @@ struct redisServer {
     /* The following two are used to track instantaneous metrics, like
      * number of operations per second, network traffic. */
     struct {
-        long long last_sample_time; /* Timestamp of last sample in ms */
-        long long last_sample_count;/* Count in last sample */
+        long long last_sample_base;  /* The divisor of last sample window */
+        long long last_sample_value; /* The dividend of last sample window */
         long long samples[STATS_METRIC_SAMPLES];
         int idx;
     } inst_metric[STATS_METRIC_COUNT];
```

```diff
--- a/src/server.c
+++ b/src/server.c
@@ -695,21 +695,16 @@ int allPersistenceDisabled(void) {
 /* ======================= Cron: called every 100 ms ======================== */

 /* Add a sample to the operations per second array of samples. */
-void trackInstantaneousMetric(int metric, long long current_reading) {
-    long long now = mstime();
-    long long t = now - server.inst_metric[metric].last_sample_time;
-    long long ops = current_reading -
-                    server.inst_metric[metric].last_sample_count;
-    long long ops_sec;
+void trackInstantaneousMetric(int metric, long long current_value, long long current_base, long long factor
+    long long base = current_base - server.inst_metric[metric].last_sample_base;
+    long long value = current_value - server.inst_metric[metric].last_sample_value;
+    long long avg = base > 0 ? (value*factor/base) : 0;

-    ops_sec = t > 0 ? (ops*1000/t) : 0;
-
-    server.inst_metric[metric].samples[server.inst_metric[metric].idx] =
-        ops_sec;
+    server.inst_metric[metric].samples[server.inst_metric[metric].idx] = avg;
     server.inst_metric[metric].idx++;
     server.inst_metric[metric].idx %= STATS_METRIC_SAMPLES;
-    server.inst_metric[metric].last_sample_time = now;
-    server.inst_metric[metric].last_sample_count = current_reading;
+    server.inst_metric[metric].last_sample_base = current_base;
+    server.inst_metric[metric].last_sample_value = current_value;
 }

 /* Return the mean of all the samples. */
@@ -1294,15 +1289,21 @@ int serverCron(struct aeEventLoop *eventLoop, long long id, void *clientData) {
         atomicGet(server.stat_net_repl_input_bytes, stat_net_repl_input_bytes);
         atomicGet(server.stat_net_repl_output_bytes, stat_net_repl_output_bytes);

-        trackInstantaneousMetric(STATS_METRIC_COMMAND,server.stat_numcommands);
+        monotime current_time = getMonotonicUs();
+        long long factor = 1000000; // us
+        trackInstantaneousMetric(STATS_METRIC_COMMAND,server.stat_numcommands,current_time,factor);
         trackInstantaneousMetric(STATS_METRIC_NET_INPUT,
-                stat_net_input_bytes + stat_net_repl_input_bytes);
+                stat_net_input_bytes + stat_net_repl_input_bytes,current_time,factor);
         trackInstantaneousMetric(STATS_METRIC_NET_OUTPUT,
-                stat_net_output_bytes + stat_net_repl_output_bytes);
+                stat_net_output_bytes + stat_net_repl_output_bytes,current_time,factor);
         trackInstantaneousMetric(STATS_METRIC_NET_INPUT_REPLICATION,
-                                 stat_net_repl_input_bytes);
+                                 stat_net_repl_input_bytes,current_time,factor);
         trackInstantaneousMetric(STATS_METRIC_NET_OUTPUT_REPLICATION,
-                                 stat_net_repl_output_bytes);
+                                 stat_net_repl_output_bytes,current_time,factor);
+        trackInstantaneousMetric(STATS_METRIC_EL_CYCLE, server.duration_stats[EL_DURATION_TYPE_EL].cnt,
+                                     current_time,factor);
+        trackInstantaneousMetric(STATS_METRIC_EL_DURATION, server.duration_stats[EL_DURATION_TYPE_EL].sum,
+                                    server.duration_stats[EL_DURATION_TYPE_EL].cnt,1);
     }
```

Reply:
sounds like a good idea.